### PR TITLE
Use new ES reader implementation in OTEL

### PR DESCRIPTION
**Pull request by pavolloffay**
Signed-off-by: Pavol Loffay <ploffay@redhat.com>

This PR switches to use new ES reader implementation in OTEL all-in-one distribution. 

Notable changes:
* the OTEL ES implements storage factory
* added parsing of response error 
* added support for archive storage writer
* fixed error for reading spans from archive storage (search after has to be with sort)

### CODE REVIEW

**Comment by pavolloffay** [comment]
@joe-elliott would you like to have a look at this one too?

**Review by joe-elliott — APPROVED** [review]
It took me awhile to get my head around these changes.  A few small questions asked, but otherwise looks good.

Primarily this seems to be the addition of a `storage.Factory` shim to make use of the `elasticchsearchexporter` and `esspanreader` packages in the otel version of the all-in-one.  This required the addition of support for archive functionality which is why it was added as well.

File: cmd/opentelemetry/app/exporter/elasticsearchexporter/index_name.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2020 The Jaeger Authors.
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/opentelemetry/app/exporter/elasticsearchexporter/index_name.go:1]: We can try to do that in a separate PR. Note that one returns a list of indices and the other one a single index.

File: cmd/opentelemetry/app/internal/esclient/client.go [hunk 2]
```diff
@@ -149,8 +149,23 @@ type MultiSearchResponse struct {
 
 // SearchResponse defines search response.
 type SearchResponse struct {
-	Hits Hits                           `json:"hits"`
-	Aggs map[string]AggregationResponse `json:"aggregations,omitempty"`
+	Hits  Hits                           `json:"hits"`
+	Aggs  map[string]AggregationResponse `json:"aggregations,omitempty"`
+	Error *SearchResponseError           `json:"error,omitempty"`
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/opentelemetry/app/internal/esclient/client.go:154]: There is an error, I will test again if that is the case for all ES versions

[the remark this answers is not in the document]
Comment (pavolloffay) [reply cmd/opentelemetry/app/internal/esclient/client.go:154]: ES
```
docker run -it --rm -e "ES_JAVA_OPTS=-Xms2g -Xmx2g" -p 9200:9200 -p 9300:9300 -e "http.host=0.0.0.0" -e "discovery.type=single-node" -e "xpack.security.enabled=false" --name=elasticsearch  docker.elastic.co/elasticsearch/elasticsearch:5.6.10

docker run -it --rm -e "ES_JAVA_OPTS=-Xms2g -Xmx2g" -p 9200:9200 -p 9300:9300 -e "http.host=0.0.0.0" -e "discovery.type=single-node" --name=elasticsearch docker.elastic.co/elasticsearch/elasticsearch-oss:6.8.4

docker run -it --rm -e "ES_JAVA_OPTS=-Xms2g -Xmx2g" -p 9200:9200 -p 9300:9300 -e "http.host=0.0.0.0" -e "discovery.type=single-node" --name=elasticsearch docker.elastic.co/elasticsearch/elasticsearch-oss:7.8.1
```

All returned the following error:
```
curl -ivX POST -H "Content-Type: application/json" http://localhost:9200/jaeger/_search\?pretty -d @query.json
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:9200...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9200 (#0)
> POST /jaeger/_search?pretty HTTP/1.1
> Host: localhost:9200
> User-Agent: curl/7.66.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 160
> 
* upload completely sent off: 160 out of 160 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
HTTP/1.1 400 Bad Request
< content-type: application/json; charset=UTF-8
content-type: application/json; charset=UTF-8
< content-length: 378
content-length: 378

< 
{
  "error" : {
    "root_cause" : [
      {
        "type" : "parsing_exception",
        "reason" : "Unknown key for a VALUE_STRING in [ignore_unavailable].",
        "line" : 1,
        "col" : 24
      }
    ],
    "type" : "parsing_exception",
    "reason" : "Unknown key for a VALUE_STRING in [ignore_unavailable].",
    "line" : 1,
    "col" : 24
  },
  "status" : 400
}
* Connection #0 to host localhost left intact
```

query.json (multi search query)
```
{"ignore_unavailable": "true"}
{"query":{"term":{"traceID":{"value":"649f0a8fe3184a3f"}}},"size":10000,"terminate_after":10000,"search_after":[1598699865301838]}
```

**Pull request merged**